File: backend/app/services/base_service.py
from app.core.errors import AppError
from app.services.database import DatabaseService


class BaseService:
    table_name = ""

    def __init__(self):
        self.db = DatabaseService()

    def get_one(self, field: str, value) -> dict:
        return self.db.fetch_one(
            f"select * from {self.table_name} where {field} = %s limit 1",
            [value],
            not_found_message=f"Registro não encontrado em {self.table_name}.",
        )

    def update_by_id(self, id_field: str, id_value, payload: dict) -> dict:
        if not payload:
            raise AppError("Nenhuma alteração informada.")

        columns = list(payload.keys())
        assignments = ", ".join(f"{column} = %s" for column in columns)
        params = [payload[column] for column in columns] + [id_value]
        return self.db.fetch_one(
            f"update {self.table_name} set {assignments} where {id_field} = %s returning *",
            params,
            not_found_message=f"Registro não encontrado em {self.table_name}.",
        )

File: backend/src/services/base-service.ts
import { AppError } from "../core/errors";
import { DatabaseService } from "./database";

export class BaseService {
  protected db: DatabaseService;
  protected tableName = "";

  constructor() {
    this.db = new DatabaseService();
  }

  protected async getOne(field: string, value: unknown) {
    return this.db.fetchOne(
      `select * from ${this.tableName} where ${field} = %s limit 1`,
      [value],
      `Registro não encontrado em ${this.tableName}.`,
    );
  }

  protected async updateById(idField: string, idValue: unknown, payload: Record<string, unknown>) {
    if (!Object.keys(payload).length) {
      throw new AppError("Nenhuma alteração informada.");
    }

    const columns = Object.keys(payload);
    const assignments = columns.map((column) => `${column} = %s`).join(", ");
    const params = [...columns.map((column) => payload[column]), idValue];
    return this.db.fetchOne(
      `update ${this.tableName} set ${assignments} where ${idField} = %s returning *`,
      params,
      `Registro não encontrado em ${this.tableName}.`,
    );
  }
}
